src/main/java/org/logstash/beats/Server.java

```diff
@@ -160,6 +177,18 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws E
         public void shutdownEventExecutor() {
             try {
                 idleExecutorGroup.shutdownGracefully().sync();
+
+                // DefaultEventExecutorGroup internally executes numbers of SingletonEventExecutor
+                // try to gracefully shut down every thread if they have unacked pending batches (pending tasks)
+                for (final EventExecutor eventExecutor : beatsHandlerExecutorGroup) {
+                    if (eventExecutor instanceof SingleThreadEventExecutor) {
+                        final SingleThreadEventExecutor singleExecutor = (SingleThreadEventExecutor) eventExecutor;
+                        if (singleExecutor.pendingTasks() > 0) {
+                            singleExecutor.shutdownGracefully().sync();
+                        }
+                    }
+                }
```

Comment:
Why we need this if the `DefaultEventExecutorGroup.shutdownGracefully` already does it, but without considering the `pendingTasks`?
As I can understand we split the shutdown of the `beatsHandlerExecutorGroup` in two phases:
- in the first we shutdown the executors that has pending tasks
- in the second the executors remained

Why we need it?
If we need it, could be better to extract this loop in a method that is more explicit of the intention, like ofr example `shutdownExecutorsWithPendingTasks` ?

https://github.com/netty/netty/blob/netty-4.1.109.Final/common/src/main/java/io/netty/util/concurrent/MultithreadEventExecutorGroup.java#L161-L166

Reply:
- Separated into `shutdownEventExecutorsWithPendingTasks` method. 
- `beatsHandlerExecutorGroup.shutdownGracefully().sync();` is a concurrent approach (we don't know which thread occupies CPU and runs). Asking each busy/pending executor to shutdown gives opportunity to process the events it has (as much as prevents data loss), especially if quiet time rollout (see https://netty.io/4.1/api/io/netty/util/concurrent/EventExecutorGroup.html#shutdownGracefully-long-long-java.util.concurrent.TimeUnit-).